### PR TITLE
Paramedic improvements

### PR DESCRIPTION
**Pull request by sgrebnov**
This PR is a replacement to original PR [Real time logging and other improvements](https://github.com/apache/cordova-paramedic/pull/1) which adds real-time logging and custom jasmine reporters support. In comparisons to original PR:
 - Removed multiple targets support
 - Refactored/simplified logic

### CODE REVIEW

**Comment by sarangan12** [comment]
LGTM

File: lib/LocalServer.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
+var Q = require('q'),
```
Comment (nikhilkh) [inline lib/LocalServer.js]: Apache license needed in comments

File: lib/paramedic.js [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        self.server.on('disconnect', function() {
+            reject(new Error('device is disconnected before passing the tests'));
+        });
+
+        var command = self.getCommandForStartingTests();
+        logger.normal('cordova-paramedic: running command ' + command);
+
+        exec(command, function(code, output) {
+            if(code) 
+                return reject(new Error(command + " returned error code " + code));
```
Comment (nikhilkh) [inline lib/paramedic.js]: `output` needs to be logged to help with diagnosing errors.

**Comment by nikhilkh** [comment]
Travis CI is failing for some reason. Can you also provide details of what this PR accomplishes?

File: paramedic-plugin/paramedic.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+var io = cordova.require('cordova-plugin-paramedic.socket.io');
```
Comment (dblotsky) [inline paramedic-plugin/paramedic.js]: Do we need a special copy of socket.io? Can we just use `require('socket.io')`?

**Comment by dblotsky** [comment]
+1 to Nikhil's comment: could you please provide a short description?

It looks like this code has some improvements over #1? Also, it looks like multiple target support has been removed; is that the case?

File: paramedic-plugin/paramedic.js (hunk 3 above)
Comment (sgrebnov) [inline paramedic-plugin/paramedic.js]: This is cordovaRequire (paramedic.js is plugin script, not a node module ) which works slightly different.
The _require("socket.io")_  and _require("./socket.io")_ won't work in this case.

File: lib/paramedic.js (hunk 2 above)
Comment (sgrebnov) [inline lib/paramedic.js]: Updated, thx!

File: lib/LocalServer.js (hunk 1 above)
Comment (sgrebnov) [inline lib/LocalServer.js]: Done

**Comment by sgrebnov** [comment]
@dblotsky, @nikhilkh - addressed CR notes, thx for review. Will update description soon. 
TODO: investigate iOS simulator deploy issue (travis-ci failure).

File: lib/LocalServer.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            });
+        }, Q(true));
+    });
+};
+
+LocalServer.prototype.createTunnel = function() {
+    logger.info('cordova-paramedic: attempt to create local tunnel');
+    var self = this;
+
+    //TODO: use localtunnel module instead of shell
```
Comment (dblotsky) [inline lib/LocalServer.js]: Is this a big change? Would it be feasible to resolve it in this PR?

File: lib/LocalServer.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            logger.info('cordova-paramedic: using tunneled url ' + tunneledUrl);
+            resolve();
+        });
+    });
+};
+
+LocalServer.prototype.createSocketListener = function() {
+    var listener = io.listen(this.port, {
+        pingTimeout: 60000, // how many ms without a pong packet to consider the connection closed
+        pingInterval: 25000 // how many ms before sending a new ping packet
```
Comment (dblotsky) [inline lib/LocalServer.js]: Please factor these values out into comments.

File: lib/LocalServer.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    var connectionUrl;
+
+    switch(platformId) {
+             case "android" : connectionUrl = "http://10.0.2.2:";
+                              break;
+             case "ios"     :
+             case "browser" :
+             case "windows" :
+             /* falls through */
+             default: connectionUrl = "http://127.0.0.1:";
```
Comment (dblotsky) [inline lib/LocalServer.js]: Please factor the URIs into constants.

File: lib/paramedic.js [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+            resolve(isTestPassed);
+        });
+
+        self.server.on('disconnect', function() {
+            reject(new Error('device is disconnected before passing the tests'));
+        });
+
+        var command = self.getCommandForStartingTests();
+        logger.normal('cordova-paramedic: running command ' + command);
```
Comment (dblotsky) [inline lib/paramedic.js]: Nitpick: might be less error-prone to factor out the logging into a function that automatically adds the `"cordova-paramedic: "` part of the string so you don't have to repeat it every time.

File: lib/utils/execWrapper.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    } else {
+        return shelljs.exec(cmd, {silent: !verbose});
+    }
+}
+
+exec.setVerboseLevel = function(_verbose) {
+    verbose = _verbose;
+};
+
+module.exports = exec;
```
Comment (dblotsky) [inline lib/utils/execWrapper.js]: What purpose does the wrapper serve?

File: paramedic-plugin/paramedic.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+new Paramedic().initialize();
+
+function loadParamedicServerUrl() {
+
+    try {
+        // attempt to synchronously load medic config
+        var xhr = new XMLHttpRequest();
+        xhr.open("GET", "../medic.json", false);
```
Comment (dblotsky) [inline paramedic-plugin/paramedic.js]: Please factor out the string into a constant.

File: paramedic-plugin/paramedic.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        xhr.send(null);
+        var cfg = JSON.parse(xhr.responseText);
+
+        return cfg.logurl || PARAMEDIC_SERVER_DEFAULT_URL;
+
+    } catch (ex) {}
```
Comment (dblotsky) [inline paramedic-plugin/paramedic.js]: Please log the error here.

File: lib/LocalServer.js [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+            var localServer = new LocalServer(port, externalServerUrl);
+            localServer.createSocketListener();
+
+            if (useTunnel) {
+                return Q.promise(function(resolve) {
+                    localServer.createTunnel().then(function(){
+                        resolve(localServer);
+                    });
+                });
```
Comment (dblotsky) [inline lib/LocalServer.js]: You can just return `localServer.createTunnel()` here.

File: lib/LocalServer.js [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+            }
+
+            return localServer;
+        });
+};
+
+LocalServer.getFirstAvailablePort = function(startPort, endPort) {
+    var ports = Array.apply(null, Array(endPort - startPort + 1)).map(function(element, index) {
+        return startPort + index;
+    });
```
Comment (dblotsky) [inline lib/LocalServer.js]: Nitpick: it would be more readable to define a `range` function that returns an array of numbers from `start` to `end` and to have `var ports = range(startPort, endPort + 1)`.

File: lib/LocalServer.js [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+            socket.on(route, function(data) {
+                self.emit(route, data);
+            });
+        });
+    });
+};
+
+LocalServer.prototype.getConnectionUrl = function() {
+    return this.tunneledUrl || 
+           (this.externalServerUrl ? this.externalServerUrl + ":" + this.port : undefined);
```
Comment (dblotsky) [inline lib/LocalServer.js]: It might be more readable to refactor this code to use if-statements instead.

**Comment by sgrebnov** [comment]
Travis-CI issue: master branch fails with the same error as well ([link](https://travis-ci.org/apache/cordova-paramedic/builds/119610330)), investigating..

File: lib/paramedic.js (hunk 7 above)
Comment (sgrebnov) [inline lib/paramedic.js]: Propose to leave as-is - we have several methods we will need to override (info, debug, normal, ..) and also several prefixes ('cordova-paramedic', 'local-server')
PS. This is out of scope of this PR as it does not aim to change originally implemented logging approach, we can create another PR to handle this.

File: lib/utils/execWrapper.js (hunk 8 above)
Comment (sgrebnov) [inline lib/utils/execWrapper.js]: It simplifies working with shell

File: paramedic-plugin/paramedic.js (hunk 9 above)
Comment (sgrebnov) [inline paramedic-plugin/paramedic.js]: Propose to leave as-is as the chance someone will want/need to change this is very low.  We should better to get rid of medic.json file usage and pass parameters as query params.

Medic.json is mostly used here to auto start tests. I believe mobile spec and test framework should only run tests and should know nothing about different target usages like medic/etc. So I think we should remove medic related functionality from test framework and add support for special query params like `autoStartTests`, etc

After above is done we should also update paramedic to pass server url as query param as well.

File: paramedic-plugin/paramedic.js (hunk 10 above)
Comment (sgrebnov) [inline paramedic-plugin/paramedic.js]: Updated, thx!

File: lib/LocalServer.js (hunk 11 above)
Comment (sgrebnov) [inline lib/LocalServer.js]: Updated, thx

File: lib/LocalServer.js (hunk 12 above)
Comment (sgrebnov) [inline lib/LocalServer.js]: Added inline comment to make this code more readable/understandable, thx

File: lib/LocalServer.js (hunk 13 above)
Comment (sgrebnov) [inline lib/LocalServer.js]: Refactored, thx!

**Comment by sgrebnov** [comment]
All CR notes have been addressed, Travis is green, Appveyour's failure is expected as there is no support to run windows app

File: lib/LocalServer.js (hunk 5 above)
Comment (sgrebnov) [inline lib/LocalServer.js]: Updated

File: lib/paramedic.js (hunk 7 above)
Comment (dblotsky) [inline lib/paramedic.js]: Sounds good to me. Thanks!

File: paramedic-plugin/paramedic.js (hunk 9 above)
Comment (dblotsky) [inline paramedic-plugin/paramedic.js]: Alright, sounds good. I'm all for not using this file or mobilespec in the future as well.

File: lib/LocalServer.js [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+                if (!isPortUsed) {
+                    resolve(port - 1);
+                    return false;
+                } else {
+                    return portChecker.check(port);
+                }
+            });
+        }, Q(true));
+    });
+};
```
Comment (dblotsky) [inline lib/LocalServer.js]: Nitpitck: the use of `reduce` here makes this code pretty hard to follow. It could be refactored to be recursive:

```javascript
return portChecker.check(startPort).then(function (inUse) {
    if (inUse) {
        return getFirstAvailablePort(startPort + 1, endPort);
    }
    resolve(startPort);
});
```

File: lib/LocalServer.js [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+
+            if (useTunnel) {
+                return localServer.createTunnel();
+            }
+
+            return localServer;
+        });
+};
+
+LocalServer.getFirstAvailablePort = function(startPort, endPort) {
```
Comment (dblotsky) [inline lib/LocalServer.js]: Should this and `startServer` be `prototype` functions?

File: lib/LocalServer.js [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+LocalServer.prototype.getConnectionUrl = function(platformId) {
+    // --useTunnel option
+    if (this.tunneledUrl) {
+        return this.tunneledUrl;
+    }
+    // --externalServerUrl option / we know ip or dns name
+    if (this.externalServerUrl) {
+        return this.externalServerUrl + ":" + this.port;
+    }
+    // build connection uri for localhost based on platform
```
Comment (dblotsky) [inline lib/LocalServer.js]: Nitpick: spaces above comments.

File: lib/paramedic.js [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+};
+
+ParamedicRunner.prototype.runTests = function() {
+    var self = this;
+
+    return Q.promise(function(resolve, reject) {
+        self.server.on('jasmineDone', function(data) {
+            logger.info('cordova-paramedic: tests have been completed');
+
+            var isTestPassed = data.specResults.specFailed === 0;
```
Comment (dblotsky) [inline lib/paramedic.js]: Please surround the expression in parentheses.

**Comment by nikhilkh** [comment]
@sgrebnov - Can you please take a look at this one? Sounds like there is some simple code feedback - let's get it in after that's fixed.

File: lib/LocalServer.js (hunk 14 above)
Comment (sgrebnov) [inline lib/LocalServer.js]: +1, updated

File: lib/LocalServer.js (hunk 15 above)
Comment (sgrebnov) [inline lib/LocalServer.js]: This method creates new LocalServer instance so should be static

File: lib/paramedic.js (hunk 17 above)
Comment (sgrebnov) [inline lib/paramedic.js]: updated
